### ISSUE TITLE
Docker support

### ISSUE THREAD

**Issue by geddski** [webpack/webpack-dev-server#143]
currently when running webpack dev server inside of a docker container, the changes to the source files are not picked up by webpack, so no compilation happens. 

Using boot2docker vm on OSX.

**Comment by csaintc** [comment]
+1 having the exact same issue, would appreciate an update

**Comment by Rodeoclash** [comment]
+1

**Comment by bradbarrow** [comment]
+1

**Comment by yvann** [comment]
+1 the same for me, I can't use it.

**Comment by evandegr** [comment]
+1

**Comment by myzone** [comment]
+1

**Comment by muraken720** [comment]
+1

**Comment by joernroeder** [comment]
+1

**Comment by johnknaack** [comment]
+1

**Comment by saulshanabrook** [comment]
Use this project: https://github.com/brikis98/docker-osx-dev

It picks up file changes events

**Comment by tomchentw** [comment]
`docker-osx-dev` works great for me. Also its woking well with [piping](https://www.npmjs.com/package/piping).

**Comment by natemara** [comment]
I am having the same problem, but I feel it's for a different reason. I'm using the `--host 0.0.0.0` option to allow the server to listen on all interfaces, but when used in conjunction with the `--hot` flag, this causes problems. The hot reloader uses the provided host to create an absolute url instead of using a relative url, so all of my reloading ajax calls are going to `http://0.0.0.0:8080/socket.io/` instead of just `socket.io/`, which would work no matter what the host/port was. I'm also proxying the server behind nginx that happens to be running on the same port, but if I wasn't, then this issue would be even worse.

**Comment by xaviervia** [comment]
+1

**Comment by vectart** [comment]
+1

**Comment by xaviervia** [comment]
@natemara about your issue, I worked around that by setting the config:

```javascript
module.exports = {
  entry: {
    app: [
      "webpack-dev-server/client?:8080/",
      'webpack/hot/dev-server',
      "./src/app.coffee"
    ]
  }
}
```

...buuut, big disclaimer, it just fixed the socket.io address, not the livereload, which is still not working (webpack doesn't rebuild the files)

**Comment by fordlee404** [comment]
+1

**Comment by plouc** [comment]
+1

**Comment by svenhornberg** [comment]
+1

**Comment by jweidler** [comment]
+1

**Comment by fzx** [comment]
The problem has sovled by Mark Wolfe ( http://www.wolfe.id.au/2015/08..., I have tested it in my MacBookPro.

**Comment by fordlee404** [comment]
I fixed the problem with webpack config:

```
watchOptions: {
    poll: true
}
```

@fzx thanks!

**Comment by vyorkin** [comment]
FYI its fixed in #236

**Comment by vyorkin** [comment]
@sokra It's not related to webpack-dev-server. It's just [vboxfs that doesn't support fsevents or inotify](https://www.virtualbox.org/ticket/10660?cversion=0&cnum_hist=1).

**Comment by vyorkin** [comment]
p.s.: here is a good post about [dev env with docker on OS X](http://www.ybrikman.com/writing/2015/05/19/docker-osx-dev/). My solution is to just set `poll: true` or if polling is too slow for you use smth like [dinghy](https://github.com/codekitchen/dinghy/).

**Comment by langri-sha** [comment]
+1

**Comment by jimthedev** [comment]
Seems like this should probably be closed since the problem is with Virtual Box's (likely reasonable) decision to not pass through inotify events. The current relatively common fix is to use https://github.com/brikis98/docker-osx-dev

This is true for anyone using node, python, or any other language in docker who wants watchers. It really has little to do with Webpack. Perhaps it just needs to be documented, but there's little to do from a code standpoint.

**Comment by langri-sha** [comment]
@jimthedev, apparently it's not even related to just `vboxsf`. It is a known issue for all NFS mounts. 

Here is an excerpt from the [PM2 documentation](http://pm2.keymetrics.io/docs/usage/watch-and-restart/):
> When working with NFS devices you'll need to set usePolling: true as stated in this [chokidar issue](https://github.com/paulmillr/chokidar/issues/242).

It seems that the best course of action is to make this information more prominent by putting it in the documentation where suitable, as you suggest.

**Comment by jimthedev** [comment]
@langri-sha, the tricky part of documenting/educating on this is that polling has implications for large project structures. It isn't always the best option. In some scenarios it might be ideal, in others, a better option might be to use docker-osx-dev to marshall the diff'ing through alternative means while leaving your webpack config untouched.

**Comment by marlonbernardes** [comment]
Thanks @fordlee404, setting `watchOptions` also worked for me.

```js
watchOptions: {
    poll: true
}
```

**Comment by nrempel** [comment]
I believe this is an issue with docker (caused by virtual box and other virtualization tools not supporting inotify events), not webpack-dev-server. 

There is a proposed solution here: https://github.com/docker/docker/issues/18246

**Comment by jimthedev** [comment]
You are correct. This is an issue with Virtual Box.

**Comment by vinogradov** [comment]
```javascript
watchOptions: {
    poll: 1000 // <-- it's worth setting a timeout to prevent high CPU load
},
```

**Comment by yangmillstheory** [comment]
Changes are picked up by the container in [Docker Mac Beta](https://blog.docker.com/2016/03/docker-for-mac-windows-beta/).

![may-25-2016 22-17-45](https://cloud.githubusercontent.com/assets/2729079/15564387/9735793e-22c6-11e6-8bf5-151dcb5aa131.gif)

**Comment by bartvanremortele** [comment]
For anyone running into this issue that is using a Jetbrains IDE like Webstorm, Intellij, ... Try:

> Please try turning 'Safe write' option ( Settings | Appearance & Behavior | System Settings | Use "safe write" (save changes to temporary file first)) off

Took me about 4 hours to figure this out.

**Comment by mrnonz** [comment]
`watchOptions: {
    poll: true
  }`

This worked for me !!

**Comment by Rastusik** [comment]
just to recap: the solution from @bartvanremortele works better than the one from @vinogradov and @mrnonz , because there is no cpu overhead at all.

> Please try turning 'Safe write' option ( Settings | Appearance & Behavior | System Settings | Use "safe > write" (save changes to temporary file first)) off

**Comment by max-mykhailenko** [comment]
@yangmillstheory Doesn't work with docker beta 20. Can you show your webpack config. I use ST3, webpack-dev-middleware, webpack-hot-middleware and express as server.

**Comment by yangmillstheory** [comment]
@max-mykhailenko I'm on `Version 1.11.1-beta12 (build: 7528)`.

Can you try https://github.com/yangmillstheory/react-docker-app?

**Comment by max-mykhailenko** [comment]
@yangmillstheory thanks, I'll check it and write results

**Comment by aethant** [comment]
I'm running into this issue with Docker beta 20, and what I'm noticing is that it "works" inconsistently. Same code, same configs to Docker and Webpack, and it'll work once, maybe a few times, then stop. Then won't work a number of times. Not sure what's causing that.

Noticing that specific IDEs/editors are mentioned here, too, I'm wondering if vim is part of the issue. 
With the Docker+Webpack setup running, I opened Atom and attempted editing the file. No dice. Restarted Docker+Webpack, edited same file, seems to respond (worked for ~5min consistently).

Why is this (file systems aren't a topic I'm well-versed on)? Is there a way to adjust vim's settings or am I stuck with Atom for the time being?

**Comment by Rodeoclash** [comment]
@aethant Have you tried setting ":set backupcopy=yes" in Vim?

**Comment by max-mykhailenko** [comment]
@yangmillstheory @aethant Clean react-docker-app works! I upgrade it from node 4.4 to 6.3 and it still works. Today I'll try dev server in my config instead express. Maybe problem is there.

**Comment by aethant** [comment]
@Rodeoclash bingo. that seems to have worked. thanks for the suggestion!

**Comment by NickStefan** [comment]
Any other findings? @aethant 

Also having this issue. Hot loading was working well but in the past few days the hot loading on docker has been breaking about 4 times a day. `Docker-compose down && docker-compose build && docker-compose up` doesnt fix it. We've been having to kill the mac meta deamon and restart it. Vim vs atom vs sublime doesnt seem to affect it (different people at our office use one or the other)

**Comment by NickStefan** [comment]
I wonder if the mac beta GUI vs the docker quickstart teminal makes any difference

**Comment by d3viant0ne** [comment]
@NickStefan - What version of Docker are you running. I haven't had any issues with Docker / Webpack since moving to the beta on mac.

@geddski As far as `boot2docker` support goes, I don't really see the point in supporting something that was handed it's hat quite a while ago and had now been `officially deprecated` in favor of `docker-machine`. The vboxfs has countless failings in regards to shared fs performance and reliability. 
//cc @sokra ^^

@yangmillstheory - That repo also doesn't work on Beta.20 for me. I'll dig into it and see if I can figure out where it's falling down.

As an aside, I'm writing docs related to Docker & Webpack2 for webpack.io and I would appreciate any insights / issues & pitfalls in webpack/webpack.io#44 from all you docker users so we can make that portion of the docs as robust as humanly possible. I believe docker for web development has significant benefits & should have first class support in Webpack.

**Comment by d3viant0ne** [comment]
@NickStefan - There is no `Docker for Mac GUI` as of yet. It just opens the existing `Kitematic`

**Comment by NickStefan** [comment]
@d3viant0ne Not public, unfortunately, and apologies for calling it a full fledged GUI. What I meant by GUI, was the whale that gets added to the upper right (haha) versus just running the script `bash --login '/Applications/Docker/Docker Quickstart Terminal.app/Contents/Resources/Scripts/start.sh'`. Sounds like that inconsequential though. 

- docker compose:
```
version: '2'
services:
   /* lots of other services */
    studio:
        build:
            context: ./node
            dockerfile: Dockerfile-studio
        volumes:
            - ./node:/usr/src
        ports:
            - "8080:8080"
```
- a simple docker file that exposes the dev server port
- the webpack set up from survivejs: https://github.com/survivejs/react-boilerplate/blob/master/webpack.config.js

One of the reasons I'm thinking this is webpack related is that we also have a django docker image that hot loads without any problems.

**Comment by d3viant0ne** [comment]
@NickStefan As an exercise for writing the Docker / Webpack docs I have been `dockerizing` popular boilerplates people are basing their projects off of.

I'll dockerize https://github.com/survivejs/react-boilerplate and hopefully it either shines a light on an issue in webpack-X or helps you work through your issue.

**Comment by aethant** [comment]
@NickStefan for me, no. Vim, even after applying `:set backupcopy=yes`, still makes Webpack inside of a container sputter out. Atom has been working without issue all the same, however.

**Comment by d3viant0ne** [comment]
@aethant - I would assume this has to do with the buffer mechanism behind vi/vim/nvim

**Comment by aethant** [comment]
@d3viant0ne Most likely you are correct. That's what I assumed, as well. But it seems like a considerable issue, given the number of vi* users. Other issues had presented the `:set backupcopy=yes` solution as a workaround, but to no avail. Kinda bummed. Vim mode in Atom just isn't the same. :/

**Comment by d3viant0ne** [comment]
@aethant You have a public project where you run into the described behavior? As a Linux / VI / Docker guy I would be curious to tinker and see if I can find a viable resolution to the VI / VIM issue.

**Comment by aethant** [comment]
Unfortunately @d3viant0ne, it's a work project. Here's a [gist](https://gist.github.com/aethant/c298fb9dec2aaba20a47425783a07454) with my webpack config and how I'm starting webpack-dev-server inside of express. Hopefully it'll be of some use to you in investigating. If you need some of the Dockerfile/docker-compose.yml bits as well, let me know.

**Comment by d3viant0ne** [comment]
@aethant It's actually the docker portion I am most curious about. Trying to figure out the common denominator between my docker Webpack setup that I don't have problems with and the setups of people that do. Outside of those still on Virtualbox of course.

**Comment by aethant** [comment]
@d3viant0ne sorry, Friday fuzz brain on my part. I included the previously mentioned gist to contain my dev Dockerfile and docker-compose.yml file.

**Comment by Cmdv** [comment]
@aethant been looking all over for a solution like this, do you have a link to a repo was interested in what `npm run dev` did as I'm trying to sort out a very slow docker/webpack rebuild process I've inherited (40 secs!!). The only downfall is that this is also a universal app!! thanks

**Comment by aethant** [comment]
@Cmdv `npm run dev` runs my `dev` script inside `package.json`, which runs a gulp routine, including `gulp-nodemon`. I can't say I have ~40s build times (like <2s-ish), but I don't know the size of your project, either.

**Comment by Cmdv** [comment]
@aethant cool thank you, could you pop up the gulp file too? just wanting to get an idea of someone's gulp file. This project is using grunt Watch and `webpack.optimize.CommonsChunkPlugin` so I think this is doing a shed load of IO which is slowing stuff loads!!! (sorry not really issue related)

**Comment by d3viant0ne** [comment]
@Cmdv - I'm saying this `as a gulp guy` 

When working within docker / shared file systems you are already going to pay a performance price. When you add gulp, which has to read more into memory from that file system for every action, that price goes up significantly.

I've cut my development cycle down considerably by limiting my gulp to utils outside of `webpack` and used straight npm scripts or @kentcdodds [p-s](https://github.com/kentcdodds/p-s) when I wanted/needed to get the bloat out of my `package.json` or needed a bit more horsepower than an `npm script` could provide.

**Comment by Cmdv** [comment]
@d3viant0ne thanks for the advice, that is normally how I go about projects and just have npm run my scripts. Think in this situation it's a combination that the project is using `grunt watch` and I'm not able to get webpack-dev-server to play nice with docker and this universal app due to the serverside rendering and webpack is doing chunks.

Adding `minChunks:` inside the `webpack.optimize.CommonsChunkPlugin({.......)}`
and `cacheDirectory` to the `loaders` helped things too.

```
{
                            test: /\.js?$/, loader: 'babel', exclude: /node_modules/,
                            query: {
                                cacheDirectory: true,
                                presets: ['babel-preset-stage-3', 'babel-preset-react', 'babel-preset-es2015']
                            }
                        },
```

Because right now the app's html is serverside driven I'm not able to set an `index.html` I assume it would require some middleware in the `express/server` part of the app.

Kind of enjoy inheriting crazy apps and thinking it might just be quicker to rebuild it all 😂

**Comment by langri-sha** [comment]
We now have access to [`CHOKIDAR_USEPOLLING`](https://github.com/paulmillr/chokidar/commit/a9d36600f11d9be7e34b4e9d489c3a14bc178514).

Run your containers as such:

```
docker run $opts -e \"CHOKIDAR_USEPOLLING=1\" $image $command
```

**Comment by d3viant0ne** [comment]
@Cmdv - Given the rise of `Universal` in Angular2 and it already being a thing in React , I'll make a note to figure this out & get it into the `HowTo`. 

Thus far professionally, the answer to issues with getting the `dev server & docker ` to play well together has always been `nginx`.

What's the rest of your stack composition look like in your `docker-compose.yml` for dev?

**Comment by Cmdv** [comment]
@d3viant0ne  Ah cool, personally I've kept away from the whole universal apps as I still think it's still too adhock but that's just me! 😋 

But notes in the `how to` would be amazing been going round in circles for a good week now!!
JavaScript, it's always the setup that takes the longest haha

```
proxy:
  image: jwilder/nginx-proxy
  environment:
    - SERVICE_IGNORE=true
  ports:
    - "8080:80"
    - "44300:443"
  volumes:
    - "/var/run/docker.sock:/tmp/docker.sock:ro"

creativepreview:
  image: docker2.******/creativepreview/app:latest
  links:
    - capiweb:capi.local
    - redis:redis.local
    - mongo:mongo.local
    - filesizeweb:filesize.local
  expose:
    - "8005"
  ports:
    - "8005:8005"
  environment:
    - ENVIRONMENT=dockerdev
    - VIRTUAL_HOST=creativepreview.local
    - VIRTUAL_PORT=8005
    - AWS_ACCESS_KEY_ID
    - AWS_REGION=eu-west-1
    - AWS_SECRET_ACCESS_KEY
  volumes:
    - /var/run/docker.sock:/tmp/docker.sock
    - ../../../:/srv/****
  command: bash -c 'grunt dev'
```

Then it runs the other containers.

**Comment by d3viant0ne** [comment]
Tagging the above as a reminder: webpack/webpack.io#44

**Comment by SpaceK33z** [comment]
Closing this because it is a solved problem.

TL;DR - Use `--watch-poll` or `watchOptions: { poll: true }` if watching doesn't work correctly. In the upcoming docs overhaul, there will be a section dedicated to Docker.

**Comment by max-mykhailenko** [comment]
With new docker all works fine, My problem was in using `volumes_from` instead of `volumes` in `docker-compose.yaml` for node container

**Comment by enjikaka** [comment]
I'm still having issues with webpack-dev-server in docker with docker-compose and nginx-proxy.

**Comment by SpaceK33z** [comment]
@enjikaka: you need to add `--host 0.0.0.0` to your `webpack-dev-server` command. By default it only allows connections from localhost.

**Comment by enjikaka** [comment]
@SpaceK33z Thanks, that made it work!

**Comment by nickdima** [comment]
Quick question for people using webpack in docker for development. Isn't it slow for non trivial apps with lots of files? After all it's running inside a VM.

**Comment by enjikaka** [comment]
@nickdima We use webpack inside Docker for deving at TIDAL. I've only tried in on my Windows machine so fat though. Works great there. :)

**Comment by bigmassa** [comment]
Not sure if this is related but I have been having a similar issue with running webpack and docker.
Scenario is on osx running `webpack --watch` from host to build my angular project which is linked in via a volume regularly crashes the container when it detects a file change. Similarly on ubuntu the container does not crash but the webpack --watch fails to recognise there is a change (which led me here).

I have found that increasing the number of file watchers in both os's have totally fixed my issues. For ubuntu by increasing the amount of inotify watchers https://github.com/guard/listen/wiki/Increasing-the-amount-of-inotify-watchers and for osx ive been running `sudo webpack --watch`.

Again this may not help but may help but thought it was a good idea to post anyway.

**Comment by evfreethy** [comment]
Anybody using RubyMine or a JetBrains IDE - please see @bartvanremortele comment to fix this issue

**Comment by jackbraj** [comment]
I also have the same problem using laravel-mix which runs webpack.
I am using it inside docker container and my editor is phpstorm.

**Comment by Angelinsky7** [comment]
For those interested (on Window), i'm currently working on a solution inspired by another....
You can try it at https://github.com/Angelinsky7/Docker-Volume-Watcher/releases and let me know.
it use a filewrapper and send notification to the docker volume container

**Comment by patroza** [comment]
@Angelinsky7 you're a genius! thanks a lot!

**Comment by Angelinsky7** [comment]
@patroza :-)

**Comment by cristianfraser** [comment]
@SpaceK33z Was a Docker section added with the docs overhaul that I'm missing or not?
